pkg/splunk/enterprise/configuration.go

```diff
@@ -778,76 +783,125 @@ func AreRemoteVolumeKeysChanged(client splcommon.ControllerClient, cr splcommon.
 }
 
-// ValidateAppFrameworkSpec checks and validates the Apps Frame Work config
-func ValidateAppFrameworkSpec(appFramework *enterprisev1.AppFrameworkSpec) error {
+func validateSplunkAppSources(appFramework *enterprisev1.AppFrameworkSpec, localScope bool) error {
 
-	// In the initial phases of App Framework feature, we are providing
-	// a flag to toggle this feature.
-	if !appFramework.FeatureEnabled {
-		return nil
-	}
+	duplicateChecker := make(map[string]bool)
+	var vol string
 
-	//log all the inputs for AppFrameworkSpec.
-	//TBD: Future - use logging toggle switch to reduce any excessive logging.
-	scopedLog := log.WithName("ValidateAppFrameworkSpec").WithValues(
-		"FeatureEnabled", appFramework.FeatureEnabled,
-		"Type", appFramework.Type,
-		"S3Endpoint", appFramework.S3Endpoint,
-		"S3Bucket", appFramework.S3Bucket,
-		"S3SecretRef", appFramework.S3SecretRef,
-		"S3PollInterval", appFramework.S3PollInterval)
+	// Make sure that all the App Sources are provided with the mandatory config values.
+	for i, appSrc := range appFramework.AppSources {
+		if appSrc.Name == "" {
+			return fmt.Errorf("App Source name is missing for AppSource at: %d", i)
+		}
 
-	scopedLog.Info("Validating app framework spec")
+		if appSrc.Location == "" {
+			return fmt.Errorf("App Source location is missing for AppSource: %s", appSrc.Name)
+		}
 
-	if appFramework.Type == "" {
-		return fmt.Errorf("Apps Remote Storage Type is missing")
-	}
+		if appSrc.VolName != "" {
+			_, err := checkIfVolumeExists(appFramework.VolList, appSrc.VolName)
+			if err != nil {
+				return fmt.Errorf("Invalid Volume Name for App Source: %s. %s", appSrc.Name, err)
+			}
+			vol = appSrc.VolName
+		} else {
+			if appFramework.Defaults.VolName == "" {
+				return fmt.Errorf("volumeName is missing for App Source: %s", appSrc.Name)
+			}
+			vol = appFramework.Defaults.VolName
+		}
 
-	if strings.ToLower(appFramework.Type) != "s3" {
-		return fmt.Errorf("Currently supported type for Apps Remote Storage Type is s3 only")
-	}
+		if appSrc.Scope != "" {
+			if localScope && appSrc.Scope != "local" {
+				return fmt.Errorf("Invalid scope for App Source: %s. Only local scope is supported for this kind of CR", appSrc.Name)
+			}
+
+			if appSrc.Scope != "local" && appSrc.Scope != "cluster" {
+				return fmt.Errorf("Scope for App Source: %s should be either local Or cluster", appSrc.Name)
+			}
+		}
+
+		if _, ok := duplicateChecker[vol+appSrc.Location]; ok {
+			return fmt.Errorf("Duplicate App Source configured for Volume: %s, and Location: %s combo. Remove the duplicate entry and reapply the configuration", vol, appSrc.Location)
+		}
+		duplicateChecker[vol+appSrc.Location] = true
 
-	if appFramework.S3Endpoint == "" {
-		return fmt.Errorf("Apps Remote Storage S3Endpoint is missing")
 	}
 
-	if appFramework.S3Bucket == "" {
-		return fmt.Errorf("Apps Remote Storage S3Bucket is missing")
+	if localScope && appFramework.Defaults.Scope != "" && appFramework.Defaults.Scope != "local" {
```

Comment:
What happens if Defaults.Scope = "" and scope has not been explicitly specified? Should we default to local/cluster based on the CR?

Reply:
Good catch. adding extra check for returning error in that case.

Reply:
Also, added a UT to catch the same.